[dvsim/lint] Use relative links on summary page

msfschaffner opened the pull request:
Instead of absolute links, this creates relative links to link to the individual lint results from the summary page. This allows us to properly present the lint results produced with the kokoro runner in GitHub PR CI.

Note that while the shared `FlowCfg` function `_get_results_page_link` has been updated, the `SimCfg` class still needs to be updated to take advantage of this new feature.

Signed-off-by: Michael Schaffner <msf@opentitan.org>

moidx commented:
This fixes #13463

moidx commented on util/dvsim/LintCfg.py:
```diff
@@ -74,12 +74,12 @@ def gen_results_summary(self):
-                cfg.results_server_prefix, cfg.results_server_url_prefix)
-            name_with_link = f'[{cfg.name.upper()}]({results_page_url})'
+            name_with_link = cfg._get_results_page_link(
+                relative_to=summary_results_path)
```
The markdown link format is needed to linkify.

msfschaffner replied:
Hm... the `_get_results_page_link` function is supposed to return a markdown link if the `--publish` switch is on. Do I understand correctly that this does not use the upload functionality of Dvsim to push to the GCS bucket?

msfschaffner replied:
Ah never mind - I should have read that bug you linked first ;).

msfschaffner commented on util/dvsim/utils.py:
```diff
@@ -600,7 +599,13 @@ def clean_odirs(odir, max_odirs, ts_format=TS_FORMAT):
     if os.path.exists(odir):
         # If output directory exists, back it up.
         ts = datetime.fromtimestamp(os.stat(odir).st_ctime).strftime(ts_format)
-        shutil.move(odir, "{}_{}".format(odir, ts))
```
I am not sure this worked as intended before... i.e., the dir would be moved to a different hierarchy due to the missing parent path.

sriyerg replied:
It is used in quite a few places I believe. It works as intended. Lets meet to ensure the changes work as intended for all usecases?

msfschaffner replied:
Sure we let's discuss this this week.

The reason why I think this may not have been working as intended is that the `if os.path.exists(odir):` is typically never evaluated when using time-stamps as the dir names, since they are all unique. When switchin this to `latest` this is not true anymore, and I noticed that dvsim was moving the old `local` dirs to the wrong hierarchy with the `shutil.move(odir, "{}_{}".format(odir, ts))` command, since `odir` contains the full path, but `ts` is only a name without the full path prepended to it.

sriyerg replied:
In the case of simulations, it's the `out` directory where the "latest" runs go, which would exist if the test was run previously. 

The `ts` is suffixed at the end, so appending it to a full path should still work.

msfschaffner replied:
I think you are right and this should not be needed. 
Note however that all results folders except the newest one will then be called `latest_<ts>` instead of `<ts>`, since everytime we write a new result, we have to rename the old `latest` folder first.

msfschaffner replied:
I've removed this change to minimize impact on already existing functions.

msfschaffner commented:
Ok I tried something along the lines of what we discussed - i.e., default the report output folder name to `latest` instead of a timestamp. When invoked multiple times, this moves the old `latest` dir to a folder with the timestamp as its name.

This works for all IPs, but not for top_earlgrey due to the fact that both the summary and the top_earlgrey reports get written to the same path (`top_earlgrey/lint/`). I.e., when the summary report is written, the `latest` dir with the top_earlgrey report gets moved to a timestamped folder due to the name collision, and hence the link does not work on the dashboard.

So we need to make sure that the result paths are unique if we want to continue with this change.

We could rename the summary report so that it gets written do a different folder, e.g. `top_earlgrey_batch/lint` or `earlgrey/lint`.

msfschaffner commented on hw/top_earlgrey/lint/top_earlgrey_lint_cfgs.hjson:
```diff
@@ -301,20 +301,38 @@
+             #      ]
+             # },
+             # TODO: this is not entirely lint clean yet. Uncomment once it is clean.
+             # {    name: chip_earlgrey_cw310
```
@tjaychen we can activate this once it has been cleaned up. until then, we can leave it in the separate `cfg` file.

msfschaffner commented:
I have moved the top-level and chip-level reports into different hierarchies in the latest force push to this PR. 
This ensures that there will be no collisions between toplevel and summary reports of the same batch run.

sriyerg commented on hw/top_earlgrey/lint/top_earlgrey_lint_cfgs.hjson:
```diff
@@ -290,7 +290,7 @@
                   fusesoc_core: lowrisc:systems:top_earlgrey
                   import_cfgs: ["{proj_root}/hw/lint/tools/dvsim/common_lint_cfg.hjson"]
-                  rel_path: "hw/top_earlgrey/lint/{tool}"
+                  rel_path: "hw/top_earlgrey/chiplevel/top_earlgrey/lint/{tool}"
```
Was the existing path not already unique?

msfschaffner replied:
Not entirely - the `top_earlgrey` summary report was written to the same location as the `top_earlgrey` result before.

sriyerg replied:
Right - but they were separate html files (summary.html v/s result.html). Is that no longer enough?

msfschaffner replied:
The problem I ran into is that the summary report gets written after the individual `result.html` report. Hence, the odir logic above detects the `latest` folder containing results.html and renames it to `latest_<ts>` before writing the summary. Further disambiguation would solve that issue.

sriyerg commented on hw/_index.md:
```diff
@@ -42,14 +42,14 @@ Finally, we provide the same set of information for all available [top level des
-* [DV Style lint results (nightly)](https://reports.opentitan.org/hw/top_earlgrey/dv/lint/veriblelint/latest/results.html)
-* [Synthesis results (nightly)](https://reports.opentitan.org/hw/top_earlgrey/syn/latest/results.html)
-* [CDC results (nightly)](https://reports.opentitan.org/hw/top_earlgrey/cdc/latest/results.html)
+* [Chip-level DV simulation results, with coverage (nightly)](https://reports.opentitan.org/hw/top_earlgrey/dv/latest/results.html)
+* [Chip-level Connectivity results (nightly)](https://reports.opentitan.org/hw/top_earlgrey/conn/jaspergold/latest/results.html)
```
These also use `chip_earlgrey_asic` as the DUT.

msfschaffner replied:
Would you suggest to give them all the same rel-path?

sriyerg replied:
Yes - I suggest maintaining a uniform path for all too-flows that are run on `chip_earlgrey_asic`.

msfschaffner replied:
Sounds good, I can align that.

msfschaffner replied:
I have aligned all chip-level flows to use the same path prefix (`top_earlgrey/ip/chip_earlgrey_asic`).
I am using the `ip` hierarchy since it already exists - but I am happy to set it to something else.

msfschaffner commented:
Thanks for the review, @sriyerg!

I won't be able to make all alignments today as I'm about to leave for a longer weekend. 
I will come back to this first thing Monday morning, though.

msfschaffner commented:
Note: the latest force-push removes the hardcoded `summary_with_links = True` assignment in `dvsim.py` that I added for testing. Hence, the lint dashboard in this PR will not show the links properly until we've updated the dvsim invocation in the infrastructure repository.

matutem reviewed (approved):
LGTM, but defer to Sri who is more familiar with this flow.

msfschaffner commented on hw/_index.md:
```diff
@@ -42,14 +42,14 @@ Finally, we provide the same set of information for all available [top level des
-* [CDC results (nightly)](https://reports.opentitan.org/hw/top_earlgrey/cdc/latest/results.html)
+* [Chip-level DV simulation results, with coverage (nightly)](https://reports.opentitan.org/hw/top_earlgrey/dv/latest/report.html)
+* [Chip-level Connectivity results (nightly)](https://reports.opentitan.org/hw/top_earlgrey/conn/jaspergold/latest/report.html)
+* [Chip-level AscentLint results (nightly)](https://reports.opentitan.org/hw/chip_earlgrey_asic/lint/ascentlint/latest/report.html)
```
the `chip_earlgrey_asic` hierarchy has been used before already - we just did not point these links to the correct place.

msfschaffner replied:
so we have multiple "chip-level" instances at this hierarchy level:
`top_earlgrey`
`chip_earlgrey_asic`
`chip_earlgrey_verilator`
`chip_earlgrey_cw310`

for the first one I agree - this is already done.
for the others we need a different hierarchy to push the reports to (this has already been done like this, but the dashboard was still pointing at `top_earlgrey` instead of `chip_earlgrey_asic`).

I agree that it would be best to align the report structure with the repository tree - but the problem is that the toplevel structure would have to be changed quite significantly in order to achieve this.

one example could be to restructure as follows:
```
└── hw
    └── tops
        ├── earlgrey
        │   ├── ip
        │   │   └── ast
        │   └── ip_autogen
        │       ├── top_earlgrey
        │       ├── top_earlgrey_asic
        │       ├── top_earlgrey_cw310
        │       └── top_earlgrey_verilator
        └── englishbreakfast
...
```

msfschaffner replied:
CC @tjaychen: did you discuss any folder hierarchy changes with Chris for the new multi-top repo strategy?

sriyerg replied:
Do we really need to run fpga / verilator tops through lint?

msfschaffner replied:
we currently do not lint `top_earlgrey_verilator`. for FPGA it would be useful, but we keep it in a separate config right now (so that it can be run as a one-off pass locally). as for `chip_earlgrey_asic`, that one is definitely needed.

what we could do is remove `top_earlgrey` and only run `chip_earlgrey_asic`, since it includes `top_earlgrey`. that way we could at least work around this issue for the time being.

msfschaffner replied:
I implemented this workaround so that all chip-level results share the same path stem (`hw/top_earlgrey/`).
we may want to discuss the top-level hierarchy in the future, however, to see whether we can improve the structure.

msfschaffner commented on util/dvsim/FlowCfg.py:
```diff
@@ -87,19 +87,18 @@ def __init__(self, flow_cfg_file, hjson_data, args, mk_config):
-        self.results_summary_server_html = ""
-        self.results_summary_server_page = ""
+        self.results_server_url = ""
+        self.results_html_name = ""
```
I preserved several variables here since otherwise even more refactoring would be required in `def _publish_results(self):`.

msfschaffner commented on util/dvsim/FlowCfg.py:
```diff
@@ -139,10 +138,22 @@ def __init__(self, flow_cfg_file, hjson_data, args, mk_config):
+        self.results_server_path = (self.results_server_prefix +
+                                    self.results_server + "/" +
+                                    self.rel_path)
+        self.results_server_dir = self.results_server_path + "/latest"
```
Note that this creates the same directory structure locally and for the reports server.

msfschaffner commented on util/dvsim/FlowCfg.py:
```diff
@@ -403,44 +414,45 @@ def gen_results(self, results):
-                                           relative_to)
-        return "[%s](%s)" % (link_text, results_page_url)
+        relative_link = os.path.relpath(self.results_local_page,
+                                        relative_to)
```
this will work both locally and remotely since the structure is the same.

msfschaffner commented:
I have tested both the local reports and the publishing functionality, and they seem to work.
See for example this lint run here:
https://reports.opentitan.org/hw/top_earlgrey/lint/veriblelint/summary/latest/report.html

msfschaffner commented on hw/_index.md:
```diff
@@ -42,14 +42,14 @@ Finally, we provide the same set of information for all available [top level des
-* [DV Style lint results (nightly)](https://reports.opentitan.org/hw/top_earlgrey/dv/lint/veriblelint/latest/results.html)
-* [Synthesis results (nightly)](https://reports.opentitan.org/hw/top_earlgrey/syn/latest/results.html)
-* [CDC results (nightly)](https://reports.opentitan.org/hw/top_earlgrey/cdc/latest/results.html)
+* [Chip-level DV simulation results, with coverage (nightly)](https://reports.opentitan.org/hw/top_earlgrey/dv/latest/report.html)
```
Yeah I can remove this and rename the section.
I just wanted to emphasize that this is for the chip-level...

msfschaffner replied:
Ok I have removed the chip-level prefixes here and changed the title of the section.

msfschaffner commented on util/dvsim/FlowCfg.py:
```diff
@@ -139,9 +138,19 @@ def __init__(self, flow_cfg_file, hjson_data, args, mk_config):
-        self.results_path = (Path(self.scratch_base_path) / "reports" /
-                             self.rel_path / "latest")
+        # Construct the path variables after variable expansion.
+        self.results_local_dir = (Path(self.scratch_base_path) / "reports" /
+                                  self.rel_path / "latest")
+        self.results_local_page = (self.results_local_dir /
+                                   self.results_html_name)
```
ok will do.

msfschaffner commented on util/dvsim/FlowCfg.py:
```diff
@@ -590,14 +594,11 @@ def publish_results_summary(self):
-               self.results_summary_server_page)
+        log.info("Publishing results summary to %s", self.results_server_url)
+        cmd = (self.results_server_cmd + " cp " +
+               str(self.results_local_page) + " " +
```
let me try to change this.

sriyerg replied:
Yes - that is what I meant in my first comment.

msfschaffner commented on hw/top_earlgrey/lint/top_earlgrey_lint_cfgs.hjson:
```diff
@@ -287,34 +287,17 @@
                   fusesoc_core: lowrisc:systems:chip_earlgrey_asic
                   import_cfgs: ["{proj_root}/hw/lint/tools/dvsim/common_lint_cfg.hjson"]
-                  rel_path: "hw/chip_earlgrey_asic/lint/{tool}"
+                  rel_path: "hw/top_earlgrey/lint/{tool}"
```
@sriyerg I've removed the `top_earlgrey` run and renamed the path for `chip_earlgrey_asic` so that all chip-level tool runs share the same path stem.

sriyerg commented:
Merging since the closed source counter part of this PR is already merged.
